modules/nf-core/deseq2/differential/templates/deseq_de.R

```diff
@@ -0,0 +1,338 @@
+
+################################################
+################################################
+## Run DESeq2 processes                       ##
+################################################
+################################################
+
+dds <- DESeqDataSetFromMatrix(
+    countData = round(count.table),
+    colData = sample.sheet,
+    design = as.formula(model)
+)
+
+dds <- DESeq(
+    dds,
+    test = opt\$test,
+    fitType = opt\$fit_type,
+    minReplicatesForReplace = opt\$min_replicates_for_replace,
+    useT = opt\$use_t,
+    sfType = opt\$sf_type,
+    parallel=TRUE, BPPARAM=MulticoreParam(opt\$cores)
+)
+
+comp.results <-
+    results(
+        dds,
+        lfcThreshold = opt\$lfc_threshold,
+        altHypothesis = opt\$alt_hypothesis,
+        independentFiltering = opt\$independent_filtering,
+        alpha = opt\$alpha,
+        pAdjustMethod = opt\$p_adjust_method,
+        minmu = opt\$minmu,
+        contrast = c(
+            opt\$contrast_variable,
+            c(opt\$treatment_level, opt\$reference_level)
+        )
+    )
+
+################################################
+################################################
+## Generate outputs                           ##
+################################################
+################################################
+
+prefix_parts <- c('contrast_variable', 'reference_level', 'treatment_level', 'blocking_variables')
+output_prefix = gsub(' ', '-', do.call(paste, lapply(c(prefix_parts), function(x) gsub("[^[:alnum:] ]", "_", x))))
+
+contrast.name <-
+    paste(opt\$treatment_level, opt\$reference_level, sep = "_vs_")
+cat("Saving results for ", contrast.name, " ...\n", sep = "")
+
+# Differential expression table- note very limited rounding for consistency of
+# results
+
+write.table(
+    format(data.frame(gene_id = rownames(comp.results), comp.results), nsmall = 8),
+    file = paste(output_prefix, 'deseq2.results.tsv', sep = '.'),
+    col.names = TRUE,
+    row.names = FALSE,
+    sep = '\t',
+    quote = FALSE
+)
+
+# Dispersion plot
+
+png(
+    file = paste(output_prefix, 'deseq2.dispersion.png', sep = '.'),
+    width = 600,
+    height = 600
+)
+plotDispEsts(dds)
+dev.off()
+
+# R object for other processes to use
+
+saveRDS(dds, file = paste(output_prefix, 'dds.rld.rds', sep = '.'))
+
+# Size factors
+
+sf_df = data.frame(sample = names(sizeFactors(dds)), data.frame(sizeFactors(dds)))
+colnames(sf_df) <- c('sample', 'sizeFactor')
+write.table(
+    sf_df,
+    file = paste(output_prefix, 'deseq2.sizefactors.tsv', sep = '.'),
+    col.names = TRUE,
+    row.names = FALSE,
+    sep = '\t',
+    quote = FALSE
+)
+
+# Write specified matrices
+
+write.table(
+    data.frame(gene_id=rownames(counts(dds)), counts(dds, normalized = TRUE)),
+    file = paste(output_prefix, 'normalised_counts.tsv', sep = '.'),
+    col.names = TRUE,
+    row.names = FALSE,
+    sep = '\t',
+    quote = FALSE
+)
+
+if (opt\$vs_method == 'vst'){
```

Comment:
I would just export both vst counts and rlog counts.

Reply:
Seems like we're creating a lot of excess files already, no?

That's why I had switches for the expression matrix output previously, a lot of the time people probably only care about the d/e results, and could keep the disk space burden small.

Reply:
Eh, they aren't large files in the grand scheme of things. They could be not published via `publishDir` option if people don't want them.

Reply:
Actually, I was reminded by someone yesterday that the `rlog` func can be quite time consuming to run. So I really think we should keep that off by default, but I'm adding an option to run both.